diff --git a/supervisor/options.py b/lib/python/site-packages/supervisor/options.py
index c1e52a0..8f66c57 100644
--- a/supervisor/options.py
+++ b/lib/python/site-packages/supervisor/options.py
@@ -867,6 +867,7 @@ class ServerOptions(Options):
         serverurl = get(section, 'serverurl', None)
         if serverurl and serverurl.strip().upper() == 'AUTO':
             serverurl = None
+        oom_score_adj = get(section, 'oom_score_adj', None)

         # find uid from "user" option
         user = get(section, 'user', None)
@@ -878,7 +879,6 @@ class ServerOptions(Options):
         umask = get(section, 'umask', None)
         if umask is not None:
             umask = octal_type(umask)
-
         process_name = process_or_group_name(
             get(section, 'process_name', '%(program_name)s', do_expand=False))

@@ -971,6 +971,7 @@ class ServerOptions(Options):
                 killasgroup=killasgroup,
                 exitcodes=exitcodes,
                 redirect_stderr=redirect_stderr,
+                oom_score_adj=oom_score_adj,
                 environment=environment,
                 serverurl=serverurl)

@@ -1539,6 +1540,14 @@ class ServerOptions(Options):
             if fd is not None:
                 self.close_fd(fd)

+    def set_oom_score_adj(self, oom_score_adj):
+        try:
+            procfile = open('/proc/%s/oom_score_adj' % os.getpid(), 'w')
+            procfile.write(str(oom_score_adj) + '\n')
+            procfile.close()
+        except IOError:
+            return "Can't set oom_score_adj to %s" % oom_score_adj
+
 class ClientOptions(Options):
     positional_args_allowed = 1

@@ -1743,7 +1752,7 @@ class ProcessConfig(Config):
         'stderr_logfile_backups', 'stderr_logfile_maxbytes',
         'stderr_events_enabled',
         'stopsignal', 'stopwaitsecs', 'stopasgroup', 'killasgroup',
-        'exitcodes', 'redirect_stderr' ]
+        'exitcodes', 'redirect_stderr', 'oom_score_adj' ]
     optional_param_names = [ 'environment', 'serverurl' ]

     def __init__(self, options, **params):
diff --git a/supervisor/process.py b/lib/python/site-packages/supervisor/process.py
index f9ddcd9..44fadbc 100644
--- a/supervisor/process.py
+++ b/lib/python/site-packages/supervisor/process.py
@@ -292,6 +292,13 @@ class Subprocess:
             self._prepare_child_fds()
             # sending to fd 2 will put this output in the stderr log

+            # set oom_score_adj, better to do it before dropping privileges
+            # so it can also be decreased
+            oom_score_adj_msg = self.set_oom_score_adj()
+            if oom_score_adj_msg:
+                options.write(2, "supervisor: %s\n" % oom_score_adj_msg)
+                return
+
             # set user
             setuid_msg = self.set_uid()
             if setuid_msg:
@@ -569,6 +576,11 @@ class Subprocess:
         msg = self.config.options.dropPrivileges(self.config.uid)
         return msg

+    def set_oom_score_adj(self):
+        if self.config.oom_score_adj is None:
+            return
+        return self.config.options.set_oom_score_adj(self.config.oom_score_adj)
+
     def __cmp__(self, other):
         # sort by priority
         return cmp(self.config.priority, other.config.priority)
